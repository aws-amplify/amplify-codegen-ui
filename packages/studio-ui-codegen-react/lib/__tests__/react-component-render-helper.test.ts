/*
  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.

  Licensed under the Apache License, Version 2.0 (the "License").
  You may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */
import {
  getFixedComponentPropValueExpression,
  ComponentPropertyValueTypes,
  getComponentPropName,
  isFixedPropertyWithValue,
  isBoundProperty,
  isCollectionItemBoundProperty,
  isConcatenatedProperty,
  isConditionalProperty,
} from '../react-component-render-helper';

describe('react-component-render-helper', () => {
  test('getFixedComponentPropValueExpression', () => {
    const value = 'testValue';
    expect(getFixedComponentPropValueExpression({ value }).text).toEqual(value);
  });

  describe('getComponentPropName', () => {
    test('with name', () => {
      const name = 'ComponentName';
      expect(getComponentPropName(name)).toEqual(`${name}Props`);
    });
    test('without name', () => {
      expect(getComponentPropName()).toEqual(`ComponentWithoutNameProps`);
    });
  });

  describe('property type checkers', () => {
    const propertyTypes: { [propertyType: string]: { checker: Function; property: ComponentPropertyValueTypes } } = {
      ConcatenatedStudioComponentProperty: { checker: isConcatenatedProperty, property: { concat: [] } },
      ConditionalStudioComponentProperty: {
        checker: isConditionalProperty,
        property: {
          condition: {
            property: 'user',
            field: 'age',
            operator: 'gt',
            operand: '18',
            then: {
              value: 'Vote',
            },
            else: {
              value: 'Sorry you cannot vote',
            },
          },
        },
      },
      FixedStudioComponentProperty: { checker: isFixedPropertyWithValue, property: { value: 'testValue' } },
      BoundStudioComponentProperty: {
        checker: isBoundProperty,
        property: { bindingProperties: { property: 'testBinding' } },
      },
      CollectionStudioComponentProperty: {
        checker: isCollectionItemBoundProperty,
        property: { collectionBindingProperties: { property: 'testCollectionBinding' } },
      },
    };

    Object.keys(propertyTypes).forEach((propertyType) => {
      describe(propertyType, () => {
        const {
          [propertyType]: { checker, property },
          ...otherProperties
        } = propertyTypes;
        test(`is ${propertyType}`, () => {
          expect(checker(property)).toBeTruthy();
        });

        Object.entries(otherProperties).forEach(([otherPropertyType, { property }]) => {
          test(`${otherPropertyType} is not ${propertyType}`, () => {
            expect(checker(property)).toBeFalsy();
          });
        });
      });
    });
  });
});
