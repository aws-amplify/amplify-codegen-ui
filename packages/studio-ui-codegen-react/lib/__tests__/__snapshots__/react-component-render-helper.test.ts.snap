// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`react-component-render-helper buildFixedJsxExpression boolean 1`] = `"{true}"`;

exports[`react-component-render-helper buildFixedJsxExpression number 1`] = `"{400}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed array 1`] = `"{[1, 2, 3]}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed boolean 1`] = `"{true}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed null 1`] = `"{null}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed number 1`] = `"{400}"`;

exports[`react-component-render-helper buildFixedJsxExpression parsed object 1`] = `"{{ \\"transponder\\": \\"rocinante\\" }}"`;

exports[`react-component-render-helper buildFixedJsxExpression string 1`] = `"\\"some text\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped array 1`] = `"\\"[1,2,3]\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped boolean 1`] = `"\\"true\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped null 1`] = `"\\"null\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped number 1`] = `"\\"400\\""`;

exports[`react-component-render-helper buildFixedJsxExpression string wrapped object 1`] = `"\\"{\\\\\\"transponder\\\\\\": \\\\\\"rocinante\\\\\\"}\\""`;
